internal/impl/generate.go

```diff
@@ -62,26 +64,81 @@ func generateForShell(rootPath string, plan *plansdk.ShellPlan, pluginManager *p
 	return nil
 }
 
-func writeFromTemplate(path string, plan interface{}, tmplName string) error {
-	embeddedPath := fmt.Sprintf("tmpl/%s.tmpl", tmplName)
+// Cache and buffers for generating templated files.
+var (
+	tmplCache = map[string]*template.Template{}
 
-	// Should we clear the directory so we start "fresh"?
-	outPath := filepath.Join(path, tmplName)
-	outDir := filepath.Dir(outPath)
-	err := os.MkdirAll(outDir, 0755) // Ensure directory exists.
+	// Most generated files are < 4KiB.
+	tmplNewBuf = bytes.NewBuffer(make([]byte, 0, 4096))
+	tmplOldBuf = bytes.NewBuffer(make([]byte, 0, 4096))
+)
+
+func writeFromTemplate(path string, plan any, tmplName string) error {
+	tmplKey := tmplName + ".tmpl"
+	tmpl := tmplCache[tmplKey]
+	if tmpl == nil {
+		tmpl = template.New(tmplKey)
+		tmpl.Funcs(templateFuncs)
+
+		var err error
+		tmpl, err = tmpl.ParseFS(tmplFS, "tmpl/"+tmplKey)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		tmplCache[tmplKey] = tmpl
+	}
+	tmplNewBuf.Reset()
+	if err := tmpl.Execute(tmplNewBuf, plan); err != nil {
+		return errors.WithStack(err)
+	}
+
+	// In some circumstances, Nix looks at the mod time of a file when
```

Comment:
is there any resource to read about which circumstances apply for this situation?